torch/export/export_schema.py

```diff
@@ -0,0 +1,326 @@
+from dataclasses import dataclass
+from enum import auto, Enum
+from typing import List, Union
+
+################################################################################
+# Following section is the defining a few utility data structures
+
+# As many of the serialization library doesn't have the concept of map
+# We use a list of sting-string pair to represent a map<string, string>
+@dataclass
+class StringStringPair:
+    key: str
+    value: str
+
+################################################################################
+# Following section is the defining the permissible argument types for operators
+
+# Copied from torchgen/model.py
+# !!! Consider if we want to add aliased dtypes, e.g float == float32. (Sherlock: No!)
+class ScalarType(Enum):
+    Byte = auto()
+    Char = auto()
+    Short = auto()
+    Int = auto()
+    Long = auto()
+    Half = auto()
+    Float = auto()
+    Double = auto()
+    ComplexHalf = auto()
+    ComplexFloat = auto()
+    ComplexDouble = auto()
+    Bool = auto()
+    BFloat16 = auto()
+
+# Copied from torch/_C/__init__.pyi.in
+# !!! Consider if we want to keep all of them, as we figures out what to do with SparseTensor
+class Layout(Enum):
+    # Defined in torch/csrc/utils/tensor_layouts.cpp
+    strided = auto()
+    sparse_coo = auto()
+    sparse_csr = auto()
+    sparse_csc = auto()
+    sparse_bsr = auto()
+    sparse_bsc = auto()
+    _mkldnn = auto()
+
+
+# Copied from torch/_C/__init__.pyi.in
+class MemoryFormat(Enum):
+    # Defined in torch/csrc/utils/tensor_memoryformats.cpp
+    contiguous_format = auto()
+    channels_last = auto()
+    channels_last_3d = auto()
+
+    # !!! This is annoying! preserve_format cannot be used as a tensor propery, as it can only used as an argument for operators
+    preserve_format = auto()
+
+
+# Copied from torch/_C/__init__.pyi.in
+@dataclass
+class Device:
```

Comment:
device is not needed: once on file there is no longer concept of device; on loading; user can choose where to put the tensor

Reply:
> device is not needed: once on file there is no longer concept of device; on loading; user can choose where to put the tensor

Device is a permissible type for Argument, so it's needed. 
e.g. aten.zeros([2,2], device= torch.device('cuda:0'))

Reply:
> We may want to capture the intention of the person doing export. For example, some models are heterogeneous and some part is computed on CPU, some part on GPU. There should be some way of encoding that.

fx.node doesn't have an explicit annotation on device placement. Instead, device placement is encoded in tensor. FakeTensor should have faithfully captured the device info.

Reply:
related https://github.com/pytorch/pytorch/pull/91779 (cc @voznesenskym)

Reply:
This point about "capturing original intent" is also related to the layout question. If we do this then we should then likely do consistently for layout and any other target-specific metadata... and then have a clear contract for how those fields are meant to be interpreted (i.e. just fyi annotations?).

Reply:
Is capturing original intent useful?

Take `device` as an example. A model author would likely have "training on cluster" in their mind when creating the model. This intent is encoded as `device` argument in there code. PT2 Export aims to get the abstract mathematical program definition out of a PyTorch program, which is a combination of program definition and runtime configurations. `device` seems more like a runtime configuration that we should ideally remove to produce a pure exported program. 

Say someone takes an exported program and tries to deploy it on TPU. How should they treat the `device=cuda:0` annotation?

Thoughts?

Reply:
> device seems more like a runtime configuration that we should ideally remove to produce a pure exported program

Overall I agree with this.

The uses I can think of for keeping the information along (same with layout), are:
1. Whether the infra needs it to reconstitute the program back into pythorch eager and want to maintain the original behaviour that included the device, etc.
2. As fyi/metadata for executors to optionally implement those semantics ... but I guess in that case it's not really metadata since it could be used, even if optionally. Plus I wonder if this information would always be able to be represented. E.g. say the concept of device starts getting more complicated and now we need to keep maintaining it here with marginal use. This already would be the case with layout where we would have to add, at least, the enums we use internally, which themselves are not very generic (ideally we would have a better representation for layout than what we currently use in pytorch frontend?).

So really, after writing this I feel more the only reason to do this is #1.
I'm curious if #1 is really needed, and/or whether anybody has other concrete use cases.

Reply:
From the discussion it seems that the serialization format is not actually trying to capture the in-memory state, but that a bit more portable representation. e.g. if you did `mode.to(cuda)` after loading the model should moving to `cuda` device becomes part of serialization? I suppose not.
In that case should reconstructing in-memory state be part of some 'on-load' lambdas that are applied when model is deserialized? Probably serializatoin of such lambdas, if needed, should be on top of this serialization schema.